server/http/handlers/service.go

```diff
@@ -146,3 +152,114 @@ func (s *service) Write(w http.ResponseWriter, r *http.Request) {
 
 	writeSuccessResponse(w)
 }
+
+func (s *service) WriteBulk(w http.ResponseWriter, r *http.Request) {
+	defer r.Body.Close()
+
+	w.Header().Set("Content-Type", "application/json")
+	if httpMethod := strings.ToUpper(r.Method); httpMethod != http.MethodPost {
+		writeErrorResponse(w, errRequestMustBePost)
+		return
+	}
+
+	data, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		err = fmt.Errorf("cannot read body: %v", err)
+		writeErrorResponse(w, err)
+		return
+	}
+
+	p := s.parserPool.Get()
+	// TODO(xichen): Check this is correct wrt the parser lifetime.
+	defer s.parserPool.Put(p)
+
+	// []byte is not a valid map key type so using string instead.
+	eventMapping := make(map[string][]event.Event)
+	var (
+		start int
+		stop  int
+	)
+	for stop < len(data) {
+		stop = bytes.Index(data[start:], writeBulkDelimiter)
+		if stop == -1 {
+			stop = len(data)
+		}
+
+		v, err := p.ParseBytes(data[start:stop])
```

Comment:
Hm we should refactor and try to reuse code.

Reply:
👍